### ISSUE TITLE
How to use PCNTLHeartbeatSender with this?

### ISSUE THREAD

**Issue by fredsted** [vyuldashev/laravel-queue-rabbitmq#384]
I'm wondering how I should implement `PCNTLHeartbeatSender` with this package. We use this package to run some very long-running jobs (30 minutes +), and we seem to get `AMQPChannelClosedException` exceptions after a while.

Due to PHP's single threaded nature, it seems that the only way to run long-running tasks using RabbitMQ this is to use php-amqplib's new PCNTLHeartbeatSender feature, as documented here: https://github.com/php-amqplib/php-amqplib/pull/815

I would like to use this in my own project, but I'm not sure where I should put the `new PCNTLHeartbeatSender` code. Does anyone have a pointer on this? Thank you.

**Comment by vyuldashev** [comment]
@fredsted Hello! Take a look at #385, please. What do you think?

**Comment by fredsted** [comment]
@vyuldashev Thanks so much for the quick response. It looks good to me. I tried this method and I got a segmentation fault a couple of seconds after I ran the `queue:work` console command.

It seems that the only way to make PCNTLHeartbeatSender work reliably is to use it in the job class, e.g.:

```
/** @property RabbitMQJob $job */
class SleepTest implements ShouldQueue
{
    use InteractsWithQueue;

    public function handle(): void
    {
        $sender = new PCNTLHeartbeatSender($this->job->getRabbitMQ()->getConnection());
        $sender->register();

        dump('before_sleep');
        $sleeps = 0;
        while ($sleeps < 20) {
            $sleeps++;
            dump('sleep '.$sleeps);
            sleep(1);
        }
        dump('after_sleep');

        $sender->unregister();
    }
}
```

If I initialize it somewhere else than the job class, I either get a segfault or it just doesn't work. For reference, I also tried a custom Connection class which I've used in the `queue.connections.rabbitmq.connection` config value, for example:

```
class AmqpConnection extends AMQPSocketConnection
{
    private static PCNTLHeartbeatSender $sender;

    public static function create_connection($hosts, $options = array())
    {
        $connection = parent::create_connection($hosts, $options);

        if ($connection->isConnected()) {
            self::$sender = new PCNTLHeartbeatSender($connection);
            self::$sender->register();
            dump('registered');
        }

        return $connection;
    }
}
```

But that doesn't work either; it's successfully registering, but the listener closure in `PCNTLHeartbeatListener:registerListener` never executes (I tried adding a dump() in there, which is the way I know it's working). Maybe it's because I'm developing on a Mac (perhaps signals behave differently under Linux?)

I'll be investigating some more.

**Comment by fredsted** [comment]
I think I found a solution that works, so I just need to extend my Job classes with a base class. Using Laravel's middleware, I can run the sender before the job executes, and terminate it after.

```php
abstract class BaseJob
{
    use InteractsWithQueue;

    public function middleware()
    {
        return [
            function ($job, $next) {
                $sender = new PCNTLHeartbeatSender($job->job->getRabbitMQ()->getConnection());
                $sender->register();
                $next($job);
                $sender->unregister();
            },
        ];
    }
}
```

And the job itself:

```php
class SleepTest extends BaseJob implements ShouldQueue
{
    public function handle(): void
    {
        dump('before_sleep');
        $sleeps = 0;
        while ($sleeps < 20) {
            $sleeps++;
            dump('sleep '.$sleeps);
            sleep(1);
        }
        dump('after_sleep');
    }
}
```

Which results in output along the lines of the following (if I add a `dump('hello') around line 78 of `PCNTLHeartbeatSender`):

```
% php artisan queue:work
[2020-12-10 10:17:12][9mPumJGI51yWLxf6ophOD4DJvKmjUGUR] Processing: App\Jobs\SleepTest
^ "before_sleep"
^ "sleep 1"
^ "sleep 2"
^ "hello"
^ "sleep 3"
^ "sleep 4"
^ "hello"
^ "sleep 5"
```

**Comment by adm-bome** [comment]
In the default settings we use a AmqpLazyConnection. this connection is not always open or opened when created.
The code in the PCNTLHeartbeatSender class forbids to register against a closed connection.

I added a PR #386 on-top @vyuldashev PR #385.

Maybe this feature is also a nice solution for your problem. ;)

**Comment by fredsted** [comment]
@adm-bome Indeed, I also found that using the lazy connection mechanism the signal wasn't being registered.

Tangent: Something else I found while testing this is that the ALERT signal is not always processed, even when it works (it doesn't always, as I found out in my previous post in this thread). 

For example, downloading a file from an SFTP server using Laravel's built-in mechanism, the signal is handled without issues during file download. But in the step of copying the file from a temporary folder to the actual destination, the signal is not processed, so the heartbeat fails, and the connection is closed. 

As a result, I've increased the heartbeat duration to 60 seconds, which leaves my app with 30 seconds to copy the file, which is enough for my specific use case.

Generally I think PHP is simply not ideal for these kinds of long-running, network-bound workflows.
